libs/langchain/tests/unit_tests/schema/runnable/test_history.py

```diff
@@ -0,0 +1,219 @@
+import datetime
+from typing import Callable, Sequence
+
+from langchain.memory import ChatMessageHistory
+from langchain.pydantic_v1 import BaseModel
+from langchain.schema import AIMessage, BaseMessage, HumanMessage
+from langchain.schema.runnable import RunnableConfig, RunnableLambda
+
+
+def get_factory() -> Callable:
+    chat_history_store = {}
+
+    def history_factory(session_id: str) -> ChatMessageHistory:
+        if session_id not in chat_history_store:
+            chat_history_store[session_id] = ChatMessageHistory()
+        return chat_history_store[session_id]
+
+    return history_factory
+
+
+def test_input_messages() -> None:
+    runnable = RunnableLambda(
+        lambda messages: "you said: "
+        + "\n".join(str(m.content) for m in messages if isinstance(m, HumanMessage))
+    )
+    factory = get_factory()
+    with_history = runnable.with_message_history(factory)
+    config: RunnableConfig = {"configurable": {"thread_id": datetime.datetime.now()}}
+    output = with_history.invoke([HumanMessage(content="hello")], config)
+    assert output == "you said: hello"
+    output = with_history.invoke([HumanMessage(content="good bye")], config)
+    assert output == "you said: hello\ngood bye"
+
+
+def test_input_dict() -> None:
+    runnable = RunnableLambda(
+        lambda input: "you said: "
+        + "\n".join(
+            str(m.content) for m in input["messages"] if isinstance(m, HumanMessage)
+        )
+    )
+    factory = get_factory()
+    with_history = runnable.with_message_history(factory, input_messages_key="messages")
+    config: RunnableConfig = {"configurable": {"thread_id": datetime.datetime.now()}}
+    output = with_history.invoke({"messages": [HumanMessage(content="hello")]}, config)
+    assert output == "you said: hello"
+    output = with_history.invoke(
+        {"messages": [HumanMessage(content="good bye")]}, config
+    )
+    assert output == "you said: hello\ngood bye"
+
```

Comment:
This test currently fails:

Do we want to support first argument being `str` or list of base message and necessarily a dict? (The docs in the message history runnable seem to indicate that's the case)

```python
def test_input_arg_with_history_key() -> None:
    runnable = RunnableLambda(
        lambda input: "you said: "
                      + "\n".join(
            [str(m.content) for m in input["history"] if isinstance(m, HumanMessage)]
            + [input["input"]]
        )
    )
    factory = get_factory()
    with_history = runnable.with_message_history(
        factory, message_history_key="history"
    )
    config: RunnableConfig = {"configurable": {"thread_id": datetime.datetime.now()}}
    output = with_history.invoke("hello", config)
    assert output == "you said: hello"
    output = with_history.invoke("hello", config)
    assert output == "you said: hello\ngood bye"
```

Reply:
Discussed with @baskaryan  -- it shouldn't work with str :)